src/lib.rs

```diff
@@ -57,47 +54,53 @@ use crate::{
             workflow_completion::{self, wf_activation_completion, WfActivationCompletion},
             ActivityHeartbeat, ActivityTaskCompletion,
         },
-        temporal::api::{
-            enums::v1::WorkflowTaskFailedCause,
-            workflowservice::v1::{PollActivityTaskQueueResponse, PollWorkflowTaskQueueResponse},
-        },
+        temporal::api::enums::v1::WorkflowTaskFailedCause,
     },
     protosext::ValidPollWFTQResponse,
     task_token::TaskToken,
-    workflow_tasks::{NewWfTaskOutcome, WorkflowTaskManager},
+    worker::Worker,
+    workflow::WorkflowCachingPolicy,
+    workflow_tasks::{FailedActivationOutcome, NewWfTaskOutcome, WorkflowTaskManager},
 };
 use dashmap::DashMap;
-use futures::TryFutureExt;
+use futures::{FutureExt, TryFutureExt};
 use std::{
     convert::TryInto,
     future::Future,
-    ops::Div,
     sync::{
         atomic::{AtomicBool, Ordering},
         Arc,
     },
-    time,
 };
-use tokio::sync::Notify;
+use tokio::sync::{
+    mpsc::{unbounded_channel, UnboundedReceiver},
+    Mutex, Notify,
+};
 
 /// This trait is the primary way by which language specific SDKs interact with the core SDK. It is
 /// expected that only one instance of an implementation will exist for the lifetime of the
 /// worker(s) using it.
 #[async_trait::async_trait]
 pub trait Core: Send + Sync {
+    /// Register a worker with core. Workers poll on a specific task queue, and when calling core's
+    /// poll functions, you must provide a task queue name. Returned activations will include that
+    /// task queue name as well, so that lang may route them to the appropriate place.
+    fn register_worker(&self, config: WorkerConfig);
+
     /// Ask the core for some work, returning a [WfActivation]. It is then the language SDK's
     /// responsibility to call the appropriate workflow code with the provided inputs. Blocks
     /// indefinitely until such work is available or [Core::shutdown] is called.
     ///
     /// TODO: Examples
-    async fn poll_workflow_task(&self) -> Result<WfActivation, PollWfError>;
+    async fn poll_workflow_task(&self, task_queue: &str) -> Result<WfActivation, PollWfError>;
```

Comment:
Something still feels a bit off with this API.
We register a Worker with a specific task queue and then later poll for tasks by `task_queue`.
Maybe `register_worker` should return an object with poll methods?

Reply:
So we talked about this before and I thought we decided against it, although I forget exactly why. From an encapsulation perspective I actually really like it. My only concern is that it maybe requires pretty tight integration with Rust that might not work as well in other languages., and it substantially expands the API surface (or, at least shifts things around quite a bit).

Reply:
Other potential benefit here is it can reduce lock contention in Core, but, that's not exactly a big issue.

Reply:
Talked with @mfateev and he mentioned a preference for poll not taking queue at all and *only* including it in the response. Problem with that is core can "get ahead" of lang (ex: if we're polling queues `a`, `b`, and `c`, if lang polls we might get responses for all three very quickly and need to buffer them, and lang may not poll again for some time for whatever reason). This is pretty desirable to avoid during unit testing where right now a `poll_workflow_task` call always equals exactly one request to the server if nothing is pending, which is good for being able to count expected number of server calls exactly.

Biggest advantages to doing a queue-agnostic poll are that we can always prioritize returning pending activations (instead of not returning them if they don't match the queue you asked for), and no need for a `Worker` type handle / stateful API.

I think I can figure out ways to deal with the unit testing nuisance if only including queue in response sounds good to you @bergundy

Reply:
So, I've gone ahead and pre-emptively done it this way, which we can reverse if there are strong feelings. It makes a lot of sense to me and my concerns about the unit testing being a problem turned out to not be real at all. This keeps closer to our existing poll-respond loop and overall seems to fit pretty nicely.

Reply:
How would this API work on the lang side?
Today a lang worker polls core "directly" which means the core-lang bridge simply passes on poll requests from lang to core.
With this new change a lot more work is needed in the bridge for storing the lang callbacks and buffering poll responses since we need lang to register a new callback and a poll response to be available per queue before a response can be dispatched to lang.

Reply:
Also take note that this change implies that polling is never initiated by lang which invalidates your assumption that traces always start in lang.

Reply:
Yeah, the concern about routing makes sense to me. Polling is still initiated by lang, it just fans out to available queues. There are pretty good reasons for either design, but, given the whole point of core is to make lang small, keeping that routing stuff to a minimum does seem a compelling reason. @mfateev ?

Reply:
We went with passing the queue name